kotlinx-coroutines-core/common/src/sync/Semaphore.kt

```diff
@@ -107,105 +118,205 @@ private class SemaphoreImpl(
         if (p >= 0) return // no waiters
-        resumeNextFromQueue()
+        tryToResumeFromQueue(permits)
     }
 
-    fun incPermits() = _availablePermits.getAndUpdate { cur ->
-        check(cur < permits) { "The number of released permits cannot be greater than $permits" }
-        cur + 1
+    internal fun incPermits(delta: Int = 1) = permitsBalance.getAndUpdate { cur ->
+        assert { delta >= 1 }
+        check(cur + delta <= permits) { "The number of released permits cannot be greater than $permits" }
+        cur + delta
     }
 
-    private suspend fun addToQueueAndSuspend() = suspendAtomicCancellableCoroutine<Unit> sc@ { cont ->
+    private suspend fun tryToAddToQueue(permits: Int) = suspendAtomicCancellableCoroutine<Unit> sc@{ cont ->
         val last = this.tail
-        val enqIdx = enqIdx.getAndIncrement()
-        val segment = getSegment(last, enqIdx / SEGMENT_SIZE)
-        val i = (enqIdx % SEGMENT_SIZE).toInt()
-        if (segment === null || segment.get(i) === RESUMED || !segment.cas(i, null, cont)) {
-            // already resumed
+        val enqueueId = enqIdx.getAndIncrement()
+        val segmentId = enqueueId / SEGMENT_SIZE
+        val segment = getSegment(last, segmentId)
+        if (segment == null) {
+            // The segment is already removed
+            // Probably, this is the unreachable case
             cont.resume(Unit)
-            return@sc
+        } else {
+            val slotId = (enqueueId % SEGMENT_SIZE).toInt()
+            val prevSlot = segment.slots[slotId].getAndSet(Slot(State.SUSPEND, permits, cont))
+            // The assertion is true, cause [RESUMED] can be set up only after [SUSPEND]
+            // and [CANCELLED] can be set up only in the handler, which will be added next
+            assert { prevSlot == null }
+            cont.invokeOnCancellation(CancelSemaphoreAcquisitionHandler(this, segment, slotId, permits).asHandler)
         }
-        cont.invokeOnCancellation(CancelSemaphoreAcquisitionHandler(this, segment, i).asHandler)
+        // Help to resume slots, if accumulator has permits
+        tryToResumeFromQueue(0)
     }
 
-    @Suppress("UNCHECKED_CAST")
-    internal fun resumeNextFromQueue() {
-        try_again@while (true) {
-            val first = this.head
-            val deqIdx = deqIdx.getAndIncrement()
-            val segment = getSegmentAndMoveHead(first, deqIdx / SEGMENT_SIZE) ?: continue@try_again
-            val i = (deqIdx % SEGMENT_SIZE).toInt()
-            val cont = segment.getAndSet(i, RESUMED)
-            if (cont === null) return // just resumed
-            if (cont === CANCELLED) continue@try_again
-            (cont as CancellableContinuation<Unit>).resume(Unit)
+    internal fun tryToResumeFromQueue(permits: Int) {
+        accumulator.getAndAdd(permits)  // add thread permits to common accumulator
+        var remain = accumulator.getAndSet(0) // try to take possession of all the accumulated permits at the moment
```

Comment:
^ The pair of `getAndAdd` and `getAndAdd` is not efficient. Should be replaced with one atomic CAS-loop.